kernel/rtlil.h

```diff
@@ -657,35 +665,205 @@ namespace RTLIL
+				ch = ch >> 1;
+			}
+		}
+
+		{
+			// sketchy zone
+			str_.~string();
+			(void)new ((void*)&bits_) bitvectype(std::move(new_bits));
+			tag = backing_tag::bits;
+		}
+	}
+
+	inline RTLIL::State &operator[](int index) {
+		return bits().at(index);
+	}
+	inline RTLIL::State operator[](int index) const {
+		return bits().at(index);
+	}
+	inline bitvectype::iterator begin() {
+		return bits().begin();
+	}
+	inline bitvectype::iterator end() {
+		return bits().end();
+	}
+
+	class const_bit_view {
+	private:
+		const Const& parent;
+	public:
+		const_bit_view(const Const& c) : parent(c) {}
+
+		class const_iterator {
+		private:
+			const Const& parent;
+			size_t idx;
 
-	inline int size() const { return bits.size(); }
-	inline bool empty() const { return bits.empty(); }
-	inline RTLIL::State &operator[](int index) { return bits.at(index); }
-	inline const RTLIL::State &operator[](int index) const { return bits.at(index); }
-	inline decltype(bits)::iterator begin() { return bits.begin(); }
-	inline decltype(bits)::iterator end() { return bits.end(); }
+		public:
+			using iterator_category = std::input_iterator_tag;
+			using value_type = State;
+			using difference_type = std::ptrdiff_t;
+			using pointer = const State*;
+			using reference = State;
+
+			const_iterator(const Const& c, size_t i) : parent(c), idx(i) {}
+
+			State operator*() const {
+				if (auto bv = parent.get_if_bits())
+					return (*bv)[idx];
+
+				bool bit = (parent.get_str()[idx / 8] & (1 << (7 - (idx % 8))));
+				return bit ? State::S1 : State::S0;
+			}
+			const_iterator& operator++() {
+				++idx;
+				return *this;
+			}
+			const_iterator& operator--() {
+				--idx;
+				return *this;
+			}
+			const_iterator& operator++(int) {
+				++idx;
+				return *this;
+			}
+			const_iterator& operator--(int) {
+				--idx;
+				return *this;
+			}
+			const_iterator& operator+=(int i) {
+				idx += i;
+				return *this;
+			}
+
+			const_iterator operator+(int add) {
+				return const_iterator(parent, idx + add);
+			}
+			int operator-(const const_iterator& other) {
+				return idx - other.idx;
+			}
+
+			bool operator==(const const_iterator& other) const {
+				return idx == other.idx;
+			}
+
+			bool operator!=(const const_iterator& other) const {
+				return !(*this == other);
+			}
+		};
+
+		const_iterator begin() const {
+			return const_iterator(parent, 0);
+		}
+		const_iterator end() const {
+			return const_iterator(parent, parent.size());
+		}
+		State back() const {
+			return *const_iterator(parent, parent.size() - 1);
+		}
+		size_t size() const {
+			return (size_t)parent.size();
+		}
+		State at(size_t i) const {
+			return *const_iterator(parent, i);
+		}
+		State operator[](size_t i) const {
+			return *const_iterator(parent, i);
+		}
+		std::vector<State> to_vec() const {
+			std::vector<State> v;
+			for (auto bit : *this)
+				v.push_back(bit);
+			return v;
+		}
+	};
+
+	const const_bit_view bits() const {
```

Comment:
I'm a bit worried naming the methods the same will lead to confusion. What about not having the `bits() const` method at all, and instead retain `begin()` and `end()` only? This will mean there's only one way to iterate over a Const object, which is `for (State bit : constval)`

Reply:
1. What do you mean by retain? Const doesn't have a `begin()` method on main
2. You mean we should remove the iterator that can mutate Const? I'm pretty sure there's code that relies on that

Reply:
> 1. What do you mean by retain? Const doesn't have a `begin()` method on main

Not on main but in this PR it does.

> 2\. You mean we should remove the iterator that can mutate Const? I'm pretty sure there's code that relies on that

No, the other one, basically get rid of `const_bit_view` but keep `begin()` and `end()` on the Const itself as the way of accessing what is now `const_bit_view::const_iterator`

Reply:
So there will be two ways of iterating Const. One is `(auto a : c)` and the other is `(auto a : c.bits())`. Both give you the same values, but the second one is mutable and makes performance worse? It's doable, I suppose

Reply:
Yes, that's what I am proposing